### DISCUSSION TITLE
How do reflections work?

### DISCUSSION THREAD

**Discussion by jwwalker** [google/filament#3402] [Q&A] [answered]
To do dynamic reflective surfaces with Filament, possibly including planar mirrors, does one need to do anything besides set the appropriate material properties?  And can you give a quick and dirty explanation of how it works?

Of course I know that rendering reflections can be accomplished using ray tracing, but Filament doesn't do that.  In my previous OpenGL work, I've implemented planar mirrors by doing an extra render from the point of view of a camera reflected across the mirror, so the more mirrors are in the scene, the more renders need to be done.  I'm guessing that Filament does something much smarter than that.

**Comment by romainguy** [comment, marked as answer]
Filament currently only offers built-in support for cubemap based reflections (via `IndirectLight`). We do plan on adding screen space reflections at some point. If you want to do planar reflections you can do it exactly as you described and I believe we have a sample or two showing how to do it. CC @prideout

**Reply by jwwalker** [reply]
I'm not seeing any planar reflections in the `samples` directory.

**Reply by romainguy** [reply]
I think it's in one of the Android samples.

**Reply by prideout** [reply]
I need to clean it up but I wrote a simple mirror reflection demo for the `RenderTarget` API that can be previewed in [this branch](https://github.com/google/filament/commit/2137d03d5c11aa2ae4098123a33355b94dd2a4e2).

**Reply by jwwalker** [reply]
@prideout Thanks for the new sample!  When using an OpenGL backend, it looks right, but when using Metal, the reflection is, I don't know, inside out or something.

<img width="657" alt="weird-reflection" src="https://user-images.githubusercontent.com/730426/104666393-9d53fc00-5688-11eb-99fa-92212278e771.png">

I don't understand why the code needs to maintain a separate reflected monkey renderable, rather than putting the original monkey in the scene of the render target, and rendering it from the viewpoint of the reflected camera.

**Reply by romainguy** [reply]
@bejado FYI

**Reply by prideout** [reply]
Regarding reflected camera vs reflected geometry: yes I think either technique would work.  I did it this way for the demo to show that Filament automatically flips the shading normal when the determinant of the object-to-world transform is negative, which is something we do for glTF conformance.  If you choose to reflect the camera instead, I think you would instead need to call setFrontFaceWindingInverted on the View.

**Reply by romainguy** [reply]
And looks like we're doing it wrong with Metal :)

**Reply by bejado** [reply]
Will take a look and see where we're going wrong with Metal.

**Reply by prideout** [reply]
@bejado you're off the hook, see #3433 :)

**Reply by ast3rX** [reply]
@prideout, the mirror image has a noticable color shift. In the example the background blue color is quite different. Does your mirror material inverse logic match the applied tone mapping?

**Reply by romainguy** [reply]
It's probably that tonemaping was applied twice. The view used to render the reflection should have post processing turned off.

**Reply by prideout** [reply]
@ast3rX, That's strange because the colors look the same for me, modulo the custom border effect (line 31 in `mirror.mat`).  I've attached a screenshot that has this effect removed.

@romainguy I convinced myself to keep post processing for the offscreen view, because it might need to use features like SSAO.  Let me know if that doesn't make sense.

<img width="1136" alt="Screen Shot 2021-02-04 at 9 29 02 AM" src="https://user-images.githubusercontent.com/1288904/106931429-70b16400-66cb-11eb-951c-f7bf07347235.png">

**Reply by prideout** [reply]
Apologies, there *is* a color shift.  I didn't see it in my screenshot until I used a tool.  I think this expected since the LUT isn't perfectly invertible.  For a perfect match you can disable post-processing as Romain suggested.

**Reply by romainguy** [reply]
SSAO is not part of the post processing step. Post-processing is color grading, tonemapping, dithering and bloom. You don't want to color grade/tonemap/bloom twice :p

**Reply by prideout** [reply]
I fixed the demo (#3500) but it exposes a bug, depth testing seems to be incorrect in the offscreen view.

**Reply by ast3rX** [reply]
Thanks! I'm getting this assertion which seems to confirm your assumption:

`validateDepthStencilState:4140: failed assertion MTLDepthStencilDescriptor sets depth test but MTLRenderPassDescriptor has a nil depthAttachment texture`
